File: src/test/java/org/mitallast/queue/transport/TransportBenchmark.java
package org.mitallast.queue.transport;

import com.typesafe.config.Config;
import com.typesafe.config.ConfigFactory;
import javaslang.collection.HashMap;
import javaslang.collection.Map;
import org.junit.Before;
import org.junit.Test;
import org.mitallast.queue.common.BaseQueueTest;

import java.util.concurrent.CountDownLatch;

public class TransportBenchmark extends BaseQueueTest {

    private TransportService transportService;
    private DiscoveryNode member;
    private CountDownLatch countDownLatch;

    @Override
    protected int max() {
        return 2000;
    }

    @Override
    protected Config config() throws Exception {
        Map<String, Object> config = HashMap.of(
                "rest.enabled", false,
                "raft.enabled", false,
                "blob.enabled", false
        );
        return ConfigFactory.parseMap(config.toJavaMap()).withFallback(super.config());
    }

    @Before
    public void setUp() throws Exception {
        transportService = node().injector().getInstance(TransportService.class);
        TransportServer transportServer = node().injector().getInstance(TransportServer.class);

        TransportController transportController = node().injector().getInstance(TransportController.class);
        transportController.registerMessageHandler(TestStreamable.class, this::handle);

        member = transportServer.localNode();
        transportService.connectToNode(member);
    }

    public void handle(TestStreamable streamable) {
        countDownLatch.countDown();
    }

    @Test
    public void test() throws Exception {
        for (int e = 0; e < 10; e++) {
            System.gc();
            countDownLatch = new CountDownLatch(total());
            long start = System.currentTimeMillis();
            for (int i = 0; i < total(); i++) {
                transportService.send(member, new TestStreamable(i));
            }
            countDownLatch.await();
            long end = System.currentTimeMillis();
            printQps("send", total(), start, end);
        }
    }

    @Test
    public void testConcurrent() throws Exception {
        for (int e = 0; e < 10; e++) {
            countDownLatch = new CountDownLatch(total());
            long start = System.currentTimeMillis();
            executeConcurrent((thread, concurrency) -> {
                for (int i = thread; i < total(); i += concurrency) {
                    transportService.send(member, new TestStreamable(i));
                }
            });
            countDownLatch.await();
            long end = System.currentTimeMillis();
            printQps("send concurrent", total(), start, end);
        }
    }
}

File: src/test/java/org/mitallast/queue/transport/TransportBenchmark.kt
package org.mitallast.queue.transport

import com.typesafe.config.Config
import com.typesafe.config.ConfigFactory
import javaslang.collection.HashMap
import org.junit.Before
import org.junit.Test
import org.mitallast.queue.common.BaseIntegrationTest
import org.mitallast.queue.common.BaseQueueTest
import java.util.concurrent.CountDownLatch

class TransportBenchmark : BaseQueueTest() {

    private var transportService: TransportService? = null
    private var member: DiscoveryNode = DiscoveryNode("", 0)
    private var countDownLatch: CountDownLatch? = null

    override fun max(): Int {
        return 2000
    }

    @Throws(Exception::class)
    override fun config(): Config {
        val config = HashMap.of<String, Any>(
            "rest.enabled", false,
            "raft.enabled", false,
            "blob.enabled", false
        )
        return ConfigFactory.parseMap(config.toJavaMap()).withFallback(super.config())
    }

    @Before
    @Throws(Exception::class)
    fun setUp() {
        transportService = node().injector().getInstance(TransportService::class.java)
        val transportServer = node().injector().getInstance(TransportServer::class.java)

        val transportController = node().injector().getInstance(TransportController::class.java)
        transportController.registerMessageHandler(BaseIntegrationTest.TestStreamable::class.java) { _ ->
            countDownLatch!!.countDown()
        }

        member = transportServer.localNode()
        transportService!!.connectToNode(member)
    }

    @Test
    @Throws(Exception::class)
    fun test() {
        for (e in 0..9) {
            System.gc()
            countDownLatch = CountDownLatch(total())
            val start = System.currentTimeMillis()
            for (i in 0 until total()) {
                transportService!!.send(member, BaseIntegrationTest.TestStreamable(i.toLong()))
            }
            countDownLatch!!.await()
            val end = System.currentTimeMillis()
            printQps("send", total().toLong(), start, end)
        }
    }

    @Test
    @Throws(Exception::class)
    fun testConcurrent() {
        for (e in 0..9) {
            countDownLatch = CountDownLatch(total())
            val start = System.currentTimeMillis()
            executeConcurrent { thread, concurrency ->
                var i = thread
                while (i < total()) {
                    transportService!!.send(member, BaseIntegrationTest.TestStreamable(i.toLong()))
                    i += concurrency
                }
            }
            countDownLatch!!.await()
            val end = System.currentTimeMillis()
            printQps("send concurrent", total().toLong(), start, end)
        }
    }
}
